### ISSUE TITLE
[RFC] Design and Implementation Plan for HunyuanImage 3.0 Support

### ISSUE THREAD

**Issue by SamitHuang** [verl-project/verl-omni#207]
### 1. Motivation
`HunyuanImage 3.0` is an 80B native multimodal model (13B active parameters) that unifies multimodal understanding and generation within an autoregressive (AR) and diffusion framework. It supports multiple generation modes including `t2t`, `i2t`, `t2i`, `it2i`, and a novel `t2ti` mode (think-recaption chain: AR chain-of-thought followed by diffusion generation). 

Integrating `HunyuanImage 3.0` into `verl-omni` is a critical step for scaling our RL training capabilities to state-of-the-art MoE-based unified multimodal models. It will allow us to perform end-to-end RL (e.g., GRPO for the AR reasoning phase and Flow-GRPO for the DiT phase) on a single unified architecture. 

### 2. Current Problem
Currently, `verl-omni` supports single-modality or decoupled architectures (like Qwen-Image for diffusion and Qwen3/Qwen-VL for AR). Introducing `HunyuanImage 3.0` presents new architectural and infrastructural challenges:
1. **Unified Two-Engine Pipeline (AR + DiT)**: The model is neither purely a diffusion model nor purely an AR model. It requires a unified trainer and rollout loop that can orchestrate both the AR engine (for CoT reasoning/captioning) and the DiT engine (for pixel generation) within the same RL step.
2. **Scale & MoE Architecture**: At 80B total parameters (MoE), the model exceeds the memory limits of single-GPU and potentially single-node setups without aggressive offloading. We must ensure FSDP/Megatron (for training) and `vllm-omni` (for rollout) can handle MoE routing and distributed parallelism efficiently.
3. **Sequential Modality Generation (`t2ti`)**: Supporting the `t2ti` mode requires custom rollout orchestration where the output tokens of the AR phase directly condition the diffusion phase before the final reward is calculated.

### 3. Proposed Design & Changes
Extending `verl-omni` to feature a **Unified Model Trainer and Rollout mechanism**.

1. **Unified Model Pipeline**:
   - Create a new pipeline directory (e.g., `verl_omni/pipelines/unified_model/`) that orchestrates both AR and DiT RL components.
   - For `t2ti` rollouts, the agent will query the `vllm-omni` AR engine to generate the CoT text, and immediately pass the generated text state into the `vllm-omni` DiT engine to generate the image. The joint trajectory will be evaluated by the reward function.

2. **MoE and Multi-Node Support**:
   - Leverage multi-node FSDP (or Megatron-LM) to distribute the 80B MoE model during the training phase.
   - On the `vllm-omni` rollout side, we will configure tensor parallelism (`TP=8`) to fit the replica within a single node if possible, or build upon the multi-node rollout design to support larger slices.

3. **Shared AR/DiT Plumping**:
   - Structure the AR step as a transition-level kernel (similar to `models/types/ar.py` in unirl) to cleanly separate the Markov transitions of the AR chain from the continuous time steps of the Flow-matching chain, while sharing the same underlying optimization loop.

4. **vLLM-Omni Integration**:
   - Extend `vllm-omni` to natively serve the `HunyuanImage 3.0` weight layout, handling both the LLM and the Diffusion transformer blocks efficiently during generation.

### 4. Dev Plan & Validation

**Implementation Plan:**
- [ ] **Model Definition**: Add `HunyuanImage 3.0` (Instruct) support to `verl-omni`'s model registry, handling its specific `trust_remote_code` or native loading paths.
- [ ] **Unified Pipeline**: Implement `train_unified_model.py` and configuration files (e.g., `examples/unified_model/hi3_vllmomni.yaml`) for end-to-end training.
- [ ] **Rollout Orchestration**: Update the `vllm-omni` worker to support the sequential `t2ti` rollout logic seamlessly.
- [ ] **Weight Sync**: Ensure the ZMQ IPC weight synchronization works correctly across AR and DiT blocks and handles MoE experts efficiently.

**Validation Plan:**
- **Modality Smoke Tests**: Validate `t2t`, `i2t`, `t2i`, and `t2ti` generation on a pristine `HunyuanImage-3.0-Instruct` checkpoint (single-process) to verify basic model compatibility.
- **End-to-End RL Loop**: Run a complete unified training step (FSDP-80B + AR/DiT vLLM-Omni engines -> Rollout -> PickScore Reward -> GRPO/Flow-GRPO -> Weight Sync). Verify that the rollout produces coherent images with valid rewards and that the training step computes sensible loss and gradient norms.
- **Scale Validation**: Test on a multi-node cluster (e.g., 2 nodes of 8xH800) to guarantee FSDP and vLLM-Omni scale properly without OOM issues or weight sync deadlocks.

### 5. Future TODOs
- **Expert Parallelism (EP)**: Investigate and integrate EP for both the FSDP trainer and `vllm-omni` to optimize the memory and compute of the MoE routing beyond standard Data Parallelism / Tensor Parallelism.
- **Broaden Unified Support**: Leverage this new `unified_model` architecture to eventually support other AR+DiT models (like Bagel) under the same framework.

**Comment by wuxibin89** [comment]
> Leverage multi-node FSDP (or Megatron-LM) to distribute the 80B MoE model during the training phase.

@SamitHuang I would suggest we use [veomni](https://github.com/ByteDance-Seed/VeOmni) training backend with FSDP+SP+EP support.

Related PR: https://github.com/verl-project/verl-omni/pull/104

**Comment by SamitHuang** [comment]
> > Leverage multi-node FSDP (or Megatron-LM) to distribute the 80B MoE model during the training phase.
> 
> @SamitHuang I would suggest we use [veomni](https://github.com/ByteDance-Seed/VeOmni) training backend with FSDP+SP+EP support.
> 
> Related PR: https://github.com/verl-project/verl-omni/pull/104
> 

Sound goods for the EP support. It relies on the model implementation of hunyuanimage3.0 in VeOmni training backend.

**Comment by Xenoryn** [comment]
Hi @SamitHuang, I’d be interested in working on #207.

I’d like to start with the training-side prerequisite by adding `HunyuanImage 3.0` support to `VeOmni`. The goal would be to enable joint `AR + DiT` training with `FSDP + SP + EP`.

This could lay the groundwork for a follow-up integration in `verl-omni`.

One thing I’d like to clarify first is licensing: I’m not sure whether the official HunyuanImage 3.0 license allows its implementation to be adapted for an open-source `VeOmni / verl-omni` integration. Is there a preferred reference implementation, or would you recommend checking with the HunyuanImage maintainers first?

If this direction makes sense, I’d be happy to put together a more detailed plan and break the work into smaller, reviewable milestones.

**Comment by SamitHuang** [comment]
> Hi [@SamitHuang](https://github.com/SamitHuang), I’d be interested in working on [#207](https://github.com/verl-project/verl-omni/issues/207).
> 
> I’d like to start with the training-side prerequisite by adding `HunyuanImage 3.0` support to `VeOmni`. The goal would be to enable joint `AR + DiT` training with `FSDP + SP + EP`.
> 
> This could lay the groundwork for a follow-up integration in `verl-omni`.
> 
> One thing I’d like to clarify first is licensing: I’m not sure whether the official HunyuanImage 3.0 license allows its implementation to be adapted for an open-source `VeOmni / verl-omni` integration. Is there a preferred reference implementation, or would you recommend checking with the HunyuanImage maintainers first?
> 
> If this direction makes sense, I’d be happy to put together a more detailed plan and break the work into smaller, reviewable milestones.

@Xenoryn Thanks. I check and confirm the HYI3.0 license allow adaptation for verl-omni. Please go ahead.

**Comment by Xenoryn** [comment]
Hi @SamitHuang — quick update on #207, plus the design question I'd most like your input on.

My VeOmni HI3 PR ([#956](https://github.com/ByteDance-Seed/VeOmni/pull/956)) is progressing slowly, so in parallel I validated the whole loop on verl-omni's current FSDP2 stack. **It now works end-to-end (the AR segment genuinely participates in the gradient)**, and I'll open the corresponding PRs shortly and link them here: _[1/N] multi-stage vllm-omni weight-sync fix → [2/N] HI3 model definition → [3/N] AR GRPO + DiT Flow-GRPO joint training._

Rollout already runs on vllm-omni's native HI3 serving via a two-stage deploy config; two small upstream fixes remain ([#6768](https://github.com/vllm-project/vllm-omni/issues/6768), [#6770](https://github.com/vllm-project/vllm-omni/issues/6770)).

The design question: the RFC's *Unified Pipeline* and *Shared AR/DiT Plumbing* are UniRL-native shapes — Step/Segment protocols (`models/types/ar.py`) plus a separate `train_unified_model.py` entry. To fit verl-omni's current skeleton I landed the same training semantics differently, achieving the same training-math goals as UniRL's hi3 setup:

- the AR segment's forward/backward hangs off `DiffusersFSDPEngine` via an `ARSegmentMixin` hook;
- the unified loss lives in trainer-agnostic `losses.py` — no new protocol layer, no forked training loop.

The resulting single training step looks like this:

> Yellow nodes = HI3-specific / key differences vs other diffusion pipelines.

```mermaid
flowchart TD
    subgraph R1["① Rollout · two-stage AR+DiT (hunyuan_image_3_moe deploy config)"]
        A1["Agent loop · builds the official AR prompt client-side<br/>prompt_utils t2i_think / en_unified<br/>checkpoint ships no chat template"]:::hi3
        A2["AR engine · GPU 0-3 · TP=4<br/>temperature=0.6 · logprobs=1<br/>samples the think/recaption text"]:::hi3
        A3["AR metadata surfaced via custom_output<br/>ar_generated_token_ids / logprobs / prompt_ids<br/>depends on 3 upstream vllm-omni patches"]:::hi3
        A4["DiT engine · GPU 4-7<br/>rebuilds the identical prompt prefix · runs the steps<br/>flattens all_latents / all_timesteps / all_log_probs"]
        A1 --> A2 --> A3 -->|"ar_generated_text + prompt_text + use_system_prompt"| A4
    end

    R1 --> R2["② Reward · PickScore etc. image reward<br/>sample_level_rewards[B,1]"]

    R2 --> R3
    subgraph R3["③ Advantage · reward propagation + per-segment normalization"]
        B1["propagate_image_rewards_to_ar<br/>mean-propagates the image reward to AR along lineage<br/>no separate text reward"]:::hi3
        B2["DiT segment: compute_flow_grpo_outcome_advantage<br/>grouped by AR trajectory / prompt"]
        B3["AR segment: compute_grpo_outcome_advantage<br/>grouped by prompt (GRPO)"]:::hi3
        B1 --> B3
        B2 -.-> Bx["advantages + ar_advantages written into the batch<br/>postprocess_advantage pipeline hook (no-op by default)"]:::hi3
        B3 -.-> Bx
    end

    R3 --> R4
    subgraph R4["④ Unified loss · single optimizer step (ARSegmentMixin via engine hook)"]
        C0["optimizer_zero_grad · once"]
        C1["per timestep: DiT transition kernel forward_step<br/>FlowGRPOLoss · clip 1e-4"]
        C2["AR segment extra forward/backward<br/>module(input_ids, mode='text') · rmpad · varlen FA<br/>_ar_token_loss GRPO · clip 1e-2 · coef=1.0"]:::hi3
        C3["per-term backward accumulates into the same LoRA<br/>optimizer_step · once"]
        C0 --> C1 --> C3
        C0 --> C2 --> C3
    end

    R4 --> R5["⑤ Weight sync · one LoRA push per engine<br/>copy:true (both stages TP>1) · trainer-side temporary fused-qkv split"]:::hi3

    classDef hi3 fill:#FFF6D8,stroke:#D4A72C,stroke-width:2px,color:#4A3A00;
```

I want to be honest that I'm still unconvinced this quick implementation is the right long-term shape: it avoids new abstractions today, but the seam is diffusion-specific, and it doesn't deliver the RFC's reusable unified paradigm.

Two questions:

1. Would it be acceptable to land this interim shape now and design the unified abstraction separately?
2. Or would you prefer to settle the unified structure in this RFC discussion first — whether close to `unified_model/` + `train_unified_model.py` or adapted to verl-omni's skeleton — before the HI3 PRs land?

Happy to go either way, based on your preference.

**Comment by SamitHuang** [comment]
@Xenoryn I'd suggest 1. Do not block HunyuanImage 3.0 PRs by the unified abstraction. The RFC’s UniRL-style entrypoint  (`train_unified_model.py`) does not match how verl-omni actually grows models with the existing diffusion trainer / FSDP2 engine (same pattern as BAGEL). A second trainer loopcan freeze the wrong abstraction before we have more models requiring AR + DiT GRPO in one optimizer step.

**Comment by Xenoryn** [comment]
> [@Xenoryn](https://github.com/Xenoryn) I'd suggest 1. Do not block HunyuanImage 3.0 PRs by the unified abstraction. The RFC’s UniRL-style entrypoint (`train_unified_model.py`) does not match how verl-omni actually grows models with the existing diffusion trainer / FSDP2 engine (same pattern as BAGEL). A second trainer loopcan freeze the wrong abstraction before we have more models requiring AR + DiT GRPO in one optimizer step.

@SamitHuang Thanks for the clear call. Agreed — the PR series will land in the interim shape on the existing diffusion trainer + FSDP2 engine (AR segment via engine hook + trainer-agnostic loss, same in-pipeline growth pattern as BAGEL), with no unified abstraction in the way. We'll leave the segment contract informal and revisit extracting a `protocol/entrypoint` only when a second AR+DiT model actually needs it.
